// @ionic/angular/dist/directives/control-value-accessors/boolean-value-accessor.js
import { Directive, ElementRef, HostListener } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { setIonicClasses } from './util/set-ionic-classes';
export class BooleanValueAccessor {
    constructor(element) {
        this.element = element;
        this.onChange = () => {
            /**/ 
        };
        this.onTouched = () => {
            /**/ 
        };
    }
    writeValue(value) {
        this.element.nativeElement.checked = value;
        setIonicClasses(this.element);
    }
    _handleIonChange(value) {
        this.onChange(value);
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    _handleBlurEvent() {
        this.onTouched();
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.element.nativeElement.disabled = isDisabled;
    }
}
BooleanValueAccessor.decorators = [
    { type: Directive, args: [{
                /* tslint:disable-next-line:directive-selector */
                selector: 'ion-checkbox,ion-toggle',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: BooleanValueAccessor,
                        multi: true
                    }
                ]
            },] },
];
/** @nocollapse */
BooleanValueAccessor.ctorParameters = () => [
    { type: ElementRef, },
];
BooleanValueAccessor.propDecorators = {
    "_handleIonChange": [{ type: HostListener, args: ['ionChange', ['$event.target.checked'],] },],
    "_handleBlurEvent": [{ type: HostListener, args: ['ionBlur',] },],
};
