// @ionic/angular/dist/directives/control-value-accessors/radio-value-accessor.js
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { setIonicClasses } from './util/set-ionic-classes';
export class RadioValueAccessor {
    constructor(element) {
        this.element = element;
        this.onChange = () => {
            /**/ 
        };
        this.onTouched = () => {
            /**/ 
        };
    }
    writeValue(value) {
        this.element.nativeElement.checked = this.value = value;
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    _handleIonSelect(value) {
        this.onChange(value);
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    _handleBlurEvent() {
        this.onTouched();
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    registerOnChange(fn) {
        this.onChange = () => {
            fn(this.value);
        };
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.element.nativeElement.disabled = isDisabled;
    }
}
RadioValueAccessor.decorators = [
    { type: Directive, args: [{
                /* tslint:disable-next-line:directive-selector */
                selector: 'ion-radio',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: RadioValueAccessor,
                        multi: true
                    }
                ]
            },] },
];
/** @nocollapse */
RadioValueAccessor.ctorParameters = () => [
    { type: ElementRef, },
];
RadioValueAccessor.propDecorators = {
    "value": [{ type: Input },],
    "_handleIonSelect": [{ type: HostListener, args: ['ionSelect', ['$event.target.checked'],] },],
    "_handleBlurEvent": [{ type: HostListener, args: ['ionBlur',] },],
};
