// @ionic/angular/dist/directives/control-value-accessors/text-value-accessor.js
import { Directive, ElementRef, HostListener } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { setIonicClasses } from './util/set-ionic-classes';
export class TextValueAccessor {
    constructor(element) {
        this.element = element;
        this.onChange = () => {
            /**/ 
        };
        this.onTouched = () => {
            /**/ 
        };
    }
    writeValue(value) {
        this.element.nativeElement.value = value;
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    _handleInputEvent(value) {
        this.onChange(value);
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    _handleBlurEvent() {
        this.onTouched();
        requestAnimationFrame(() => {
            setIonicClasses(this.element);
        });
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.element.nativeElement.disabled = isDisabled;
    }
}
TextValueAccessor.decorators = [
    { type: Directive, args: [{
                /* tslint:disable-next-line:directive-selector */
                selector: 'ion-input:not([type=number]),ion-textarea,ion-searchbar',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: TextValueAccessor,
                        multi: true
                    }
                ]
            },] },
];
/** @nocollapse */
TextValueAccessor.ctorParameters = () => [
    { type: ElementRef, },
];
TextValueAccessor.propDecorators = {
    "_handleInputEvent": [{ type: HostListener, args: ['input', ['$event.target.value'],] },],
    "_handleBlurEvent": [{ type: HostListener, args: ['ionBlur',] },],
};
